File: api/schemas/campaign.py
"""Pydantic schemas for campaign endpoints."""

from pydantic import BaseModel
from api.schemas.party import PartySchema
from api.schemas.election import EngineParamsInput


class ActionInput(BaseModel):
    party: str
    action_type: str
    target_lgas: list[int] | None = None
    target_azs: list[int] | None = None
    target_districts: list[str] | None = None
    target_party: str | None = None
    language: str = "english"
    parameters: dict = {}


class CrisisInput(BaseModel):
    name: str
    turn: int = 0
    affected_azs: list[int] | None = None
    affected_lgas: list[int] | None = None
    salience_shifts: dict[str, float] = {}
    valence_effects: dict[str, float] | None = None
    awareness_boost: dict[str, float] | None = None
    tau_modifier: float = 0.0
    description: str = ""


class NewCampaignRequest(BaseModel):
    parties: list[PartySchema] = []
    params: EngineParamsInput = EngineParamsInput()
    n_monte_carlo: int = 5
    seed: int | None = None
    n_turns: int = 12


class AdvanceTurnRequest(BaseModel):
    actions: list[ActionInput] = []
    crises: list[CrisisInput] = []


class PartyStatus(BaseModel):
    name: str
    pc: float
    cohesion: float
    exposure: float
    momentum: int
    momentum_direction: str
    vote_share: float
    seats: float
    awareness: float  # mean awareness across all LGAs (0-1)
    eto_score: float  # mean ETO score across all categories/zones


class CampaignStateResponse(BaseModel):
    turn: int
    n_turns: int
    phase: str
    party_statuses: list[PartyStatus]
    scandal_history: list[dict]
    poll_results: list[dict]


class TurnResultResponse(BaseModel):
    turn: int
    state: CampaignStateResponse
    national_vote_shares: dict[str, float]
    national_turnout: float
    seat_counts: dict[str, float]
    actions_resolved: list[dict]
    synergies: list[dict]
    scandals: list[dict]

File: frontend/src/api/campaign.ts
import api from './client';
import type { Party, EngineParams, ActionInput, CrisisInput } from '../types';

export interface PartyStatus {
  name: string;
  pc: number;
  cohesion: number;
  exposure: number;
  momentum: number;
  momentum_direction: string;
  vote_share: number;
  seats: number;
  awareness: number;
  eto_score: number;
}

export interface CampaignStateResponse {
  turn: number;
  n_turns: number;
  phase: string;
  party_statuses: PartyStatus[];
  scandal_history: Record<string, unknown>[];
  poll_results: Record<string, unknown>[];
}

export interface TurnResult {
  turn: number;
  state: CampaignStateResponse;
  national_vote_shares: Record<string, number>;
  national_turnout: number;
  seat_counts: Record<string, number>;
  total_seats: number;
  actions_resolved: Record<string, unknown>[];
  synergies: Record<string, unknown>[];
  scandals: Record<string, unknown>[];
}

export async function newCampaign(parties: Party[], params: Omit<EngineParams, 'n_monte_carlo' | 'seed'>,
  n_monte_carlo: number, seed: number | null, n_turns: number): Promise<CampaignStateResponse> {
  const res = await api.post('/campaign/new', { parties, params, n_monte_carlo, seed, n_turns });
  return res.data;
}

export async function getCampaignState(): Promise<CampaignStateResponse> {
  const res = await api.get('/campaign/state');
  return res.data;
}

export async function advanceTurn(actions: ActionInput[], crises: CrisisInput[]): Promise<TurnResult> {
  const res = await api.post('/campaign/advance', { actions, crises }, { timeout: 120000 });
  return res.data;
}

export async function resetCampaign(): Promise<CampaignStateResponse> {
  const res = await api.post('/campaign/reset');
  return res.data;
}

export async function getCampaignHistory(): Promise<TurnResult[]> {
  const res = await api.get('/campaign/history');
  return res.data;
}
